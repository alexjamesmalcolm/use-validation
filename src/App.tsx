import React from "react";
import useValidation from "./lib/useValidation";
import styles from "./App.module.css";

const isNumberPositive = (number: number) => {
  if (number < 0) {
    return `${number} is less than zero`;
  }
  return true;
};

const isPokemon = (name: string) => {
  if (!name) return true;
  return fetch(`https://pokeapi.co/api/v2/pokemon/${name.toLowerCase()}`, {
    headers: { "Content-Type": "application/json" },
  })
    .then((response) => {
      if (response.status === 200) return true;
      if (response.status === 404) return `${name} is not a pokemon`;
      return false;
    })
    .catch((e) => e.message);
};

const App = () => {
  const PositiveNumberValidator = useValidation(isNumberPositive);
  const DebouncedPositiveNumberValidator = useValidation(
    isNumberPositive,
    2000
  );
  const PokemonNameValidator = useValidation(isPokemon, 2000);
  return (
    <form className={styles.container}>
      <label>
        Positive Number
        <PositiveNumberValidator>
          <input max="10" type="number" required />
        </PositiveNumberValidator>
      </label>
      <label>
        Positive Number but validated after typing stops for 2000 millisecond
        <DebouncedPositiveNumberValidator>
          <input type="number" required />
        </DebouncedPositiveNumberValidator>
      </label>
      <label>
        Pokemon name
        <PokemonNameValidator>
          <input type="text" required />
        </PokemonNameValidator>
      </label>
      <label>
        Input without a wrapper to compare styling
        <input max="10" type="number" required />
      </label>
      <input type="submit" />
    </form>
  );
};

export default App;
