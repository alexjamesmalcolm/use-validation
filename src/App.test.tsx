import React from "react";
import ReactDOM from "react-dom";
import { mount, render } from "enzyme";
import App from "./App";
import { wait } from "@testing-library/react";

it("should render without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<App />, div);
});

test("snapshot", () => {
  const wrapper = render(<App />);
  expect(wrapper).toMatchSnapshot();
});

// test("typing in the positive number field", () => {
//   const wrapper = mount(<App />);
//   const event = { target: { value: "5" } };
//   wrapper.find("#positive-number").simulate("change", event);
// });

test("New validator function should be used if it changes", async () => {
  const onSubmit = jest.fn();
  const wrapper = mount(<App onSubmit={onSubmit} />);
  wrapper.find("#first-match").simulate("change", { target: { value: "100" } });
  wrapper
    .find("#second-match")
    .simulate("change", { target: { value: "100" } });
  wrapper.find("input#submit").simulate("click");
  expect(onSubmit).toHaveBeenCalledTimes(0);
  await wait(
    () => {
      wrapper.find("input#submit").simulate("click");
      // expect(onSubmit).toHaveBeenCalled();
    },
    { timeout: 1000 }
  );
});
