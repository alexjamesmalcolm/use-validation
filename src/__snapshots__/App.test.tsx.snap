// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="container"
      >
        <label>
          Positive Number
          <div
            style="display: contents;"
          >
            <input
              max="10"
              required=""
              type="number"
            />
          </div>
        </label>
        <label>
          Positive Number but validated after typing stops for 2000 millisecond
          <div
            style="display: contents;"
          >
            <input
              required=""
              type="number"
            />
          </div>
        </label>
        <label>
          Pokemon name
          <div
            style="display: contents;"
          >
            <input
              required=""
              type="text"
            />
          </div>
        </label>
        <label>
          Input without a wrapper to compare styling
          <input
            max="10"
            required=""
            type="number"
          />
        </label>
        <input
          type="submit"
        />
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="container"
    >
      <label>
        Positive Number
        <div
          style="display: contents;"
        >
          <input
            max="10"
            required=""
            type="number"
          />
        </div>
      </label>
      <label>
        Positive Number but validated after typing stops for 2000 millisecond
        <div
          style="display: contents;"
        >
          <input
            required=""
            type="number"
          />
        </div>
      </label>
      <label>
        Pokemon name
        <div
          style="display: contents;"
        >
          <input
            required=""
            type="text"
          />
        </div>
      </label>
      <label>
        Input without a wrapper to compare styling
        <input
          max="10"
          required=""
          type="number"
        />
      </label>
      <input
        type="submit"
      />
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
